xds/internal/xdsclient/xdslbregistry/converter.go

```diff
@@ -0,0 +1,170 @@
+		case "type.googleapis.com/xds.type.v3.TypedStruct":
+			tsProto := &v3.TypedStruct{}
+			if err := proto.Unmarshal(plcy.GetTypedExtensionConfig().GetTypedConfig().GetValue(), tsProto); err != nil {
+				return nil, fmt.Errorf("failed to unmarshal resource: %v", err)
+			}
+			return convertCustomPolicyV3(tsProto)
+		case "type.googleapis.com/udpa.type.v1.TypedStruct":
+			tsProto := &v1.TypedStruct{}
+			if err := proto.Unmarshal(plcy.GetTypedExtensionConfig().GetTypedConfig().GetValue(), tsProto); err != nil {
+				return nil, fmt.Errorf("failed to unmarshal resource: %v", err)
+			}
+			return convertCustomPolicyV1(tsProto)
+		}
+	}
+	return nil, fmt.Errorf("no supported policy found in policy list +%v", policy)
+}
+
+// "the registry will maintain a set of converters that are able to map
+// from the xDS LoadBalancingPolicy to the internal gRPC JSON format"
+func convertRingHash(rhCfg *v3ringhashpb.RingHash) (json.RawMessage, error) {
+	if rhCfg.GetHashFunction() != v3ringhashpb.RingHash_XX_HASH {
+		return nil, fmt.Errorf("unsupported ring_hash hash function %v", rhCfg.GetHashFunction())
+	}
+
+	var minSize, maxSize uint64 = defaultRingHashMinSize, defaultRingHashMaxSize
+	if min := rhCfg.GetMinimumRingSize(); min != nil {
+		minSize = min.GetValue()
+	}
+	if max := rhCfg.GetMaximumRingSize(); max != nil {
+		maxSize = max.GetValue()
+	}
+
+	rhLBCfg := ringhash.LBConfig{
+		MinRingSize: minSize,
+		MaxRingSize: maxSize,
+	}
+	rhLBCfgJSON, err := json.Marshal(rhLBCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling json in ring hash converter: %v", err)
+	}
+	return makeBalancerConfigJSON(ringhash.Name, rhLBCfgJSON), nil
+}
+
+func convertWrrLocality(wrrlCfg *v3wrrlocalitypb.WrrLocality, depth int) (json.RawMessage, error) {
+	epJSON, err := ConvertToServiceConfig(wrrlCfg.GetEndpointPickingPolicy(), depth+1)
+	if err != nil {
+		return nil, fmt.Errorf("error converting endpoint picking policy: %v for %+v", err, wrrlCfg)
+	}
+	wrrCfgJSON := createWRRConfig(epJSON)
+	return makeBalancerConfigJSON(wrrlocality.Name, wrrCfgJSON), nil
+}
+
+func createWRRConfig(epCfgJSON json.RawMessage) json.RawMessage {
+	return []byte(fmt.Sprintf(`{"childPolicy": %s}`, epCfgJSON))
+}
+
+// A52 defines a LeastRequest converter but grpc-go does not support least_request.
+
+func convertCustomPolicyV3(typedStruct *v3.TypedStruct) (json.RawMessage, error) {
+	return convertCustomPolicy(typedStruct.GetTypeUrl(), typedStruct.GetValue())
+}
+
+func convertCustomPolicyV1(typedStruct *v1.TypedStruct) (json.RawMessage, error) {
+	return convertCustomPolicy(typedStruct.GetTypeUrl(), typedStruct.GetValue())
+}
+
+func convertCustomPolicy(typeURL string, s *structpb.Struct) (json.RawMessage, error) {
+	// The gRPC policy name will be the "type name" part of the value of the
+	// type_url field in the TypedStruct. We get this by using the part after
+	// the last / character. Can assume a valid type_url from the control plane.
```

Comment:
Yes, I was wondering if I could make this assumption since this isn't explicitly started in gRFC, so I pinged Terry. He mentioned Java does and he thinks you can assume this statement.